src/server/storage.rs

```diff
@@ -0,0 +1,234 @@
+    // either revert this to be `&str` or make it `AsRef<str>` everywhere
+    #[instrument(level = "debug", skip(self, model_name), fields(model_name = %model_name.as_ref()))]
+    pub async fn get(
+        &self,
+        lattice_id: &str,
+        model_name: impl AsRef<str>,
+    ) -> Result<Option<(StoredManifest, u64)>> {
+        let key = model_key(lattice_id, model_name.as_ref());
+        debug!(%key, "Fetching model from storage");
+        self.store
+            .entry(&key)
+            .await
+            .map_err(|e| anyhow::anyhow!("{e:?}"))?
+            .and_then(|entry| {
+                // Skip any delete or purge operations
+                if matches!(entry.operation, Operation::Delete | Operation::Purge) {
+                    return None;
+                }
+
+                Some(
+                    serde_json::from_slice::<StoredManifest>(&entry.value)
+                        .map_err(anyhow::Error::from)
+                        .map(|m| (m, entry.revision)),
+                )
+            })
+            .transpose()
+    }
+
+    /// Updates the stored data with the given model, overwriting any existing data. The optional
+    /// `current_revision` parameter can be used to compare whether or not you're updating the model
+    /// with the latest revision
+    #[instrument(level = "debug", skip(self, model), fields(model_name = %model.name()))]
+    pub async fn set(
+        &self,
+        lattice_id: &str,
+        model: StoredManifest,
+        current_revision: Option<u64>,
+    ) -> Result<()> {
+        debug!("Storing model in storage");
+        // We need to store the model, then update the set. This is because if we update the set
+        // first and the model fails, it will look like the model exists when it actually doesn't
+        let key = model_key(lattice_id, model.name());
+        trace!(%key, "Storing manifest at key");
+        let data = serde_json::to_vec(&model).map_err(anyhow::Error::from)?;
+        if let Some(revision) = current_revision {
+            self.store.update(&key, data.into(), revision).await
+        } else {
+            self.store.put(&key, data.into()).await
+        }
+        .map_err(|e| anyhow::anyhow!("{e:?}"))?;
+
+        trace!("Adding model to set");
+        self.retry_model_update(lattice_id, ModelNameOperation::Add(model.name()))
+            .await
+    }
+
+    /// Fetches a summary of all models in the given lattice.
+    #[instrument(level = "debug", skip(self))]
+    pub async fn list(&self, lattice_id: &str) -> Result<Vec<ModelSummary>> {
+        debug!("Fetching list of models from storage");
+        let futs = self
+            .get_model_set(lattice_id)
+            .await?
+            .unwrap_or_default()
+            .0
+            .into_iter()
+            // We can't use filter map with futures, but we can use map and then filter_map it out below
```

Comment:
Did you want to add this `filter_map` or was it not necessary in the end?

Reply:
It wasn't necessary. I'll remove the comment